### ISSUE TITLE
Per-env pipeline

### ISSUE THREAD

**Issue by jvdsande** [pikapkg/pack#29]
Hi @pikapkg team!
First of all thanks for @pika/pack =)

I was wondering today if there is a way (or plans for adding a way) to set the pipeline differently according to the NODE_ENV environment variable.

My use-case for this is quite simple: 
I want to use `@pika/plugin-build-types` (or even `@pika/plugin-ts-standard-pkg`) when building for production, but wish I could fallback to using `@pika/plugin-standard-pkg` alone while in development, since the `tsc` step is quite slow.

So doing:
```NODE_ENV=development pack build```

Would trigger only `@pika/plugin-standard-pkg`, while

```pack build``` or ```NODE_ENV=production pack build``` 

Would use `@pika/plugin-ts-standard-pkg`

Is that possible at all?

On a side-note, a `pack watch` command would be awesome !

**Comment by tsaiDavid** [comment]
A watch command would be amazing!

**Comment by ryanwalters** [comment]
+1 for `watch` command!

**Comment by ryaninvents** [comment]
I think overloading `NODE_ENV` might prevent some use cases. I've got a pipeline specifying three builders. When I'm developing, I often only want to run one builder. In this case, `NODE_ENV=development` is needed for some of the Babel transforms I use, so I'd prefer to use either a different env var, or a CLI flag.

I'd envision it as a set of "alternate" pipelines:

```json
{
  "@pika/pack": {
    "pipeline": [ "... main pipeline here ..." ],
    "envPipelines": {
      "dev": [ "... define a 'dev' pipeline ..." ],
      "foobar": [ "... define pipeline to compile only the 'foobar' asset ..."]
    }
  }
}
```

This is what Babel does [with its `env` option](https://babeljs.io/docs/en/options#env). I'd picture using this as `pack build --pipeline foobar`.

It would be a bit annoying to have to copy/paste your entire pipeline into each env to edit it, but the alternative would be using some set of `includes/excludes` specifiers, which could run into edge cases and be more difficult to troubleshoot.

**Comment by jvdsande** [comment]
Totally agree, I did not think this through with the NODE_ENV entry.
Having it as a separate, pika/pack specific flag is much better indeed!

**Comment by FredKSchott** [comment]
Hey all, thanks for getting this conversation going. I've been mulling this over a bit myself over the last few days. 

It sounds like everyone here is really interested in supporting a "faster development mode". Is that a fair characteristic / is anyone interested in a multi-env pipeline for any other reason? I think that's an important point to flesh out first, because "faster development mode" could be solved with a watch command or some other feature that allowed the user to say "only run this subset of plugins this time".

For example: I once experimented with something like "--target node" which let plugins say which targets they built for, and then only run those plugins when the `--target` arg was passed. But that got tricky when some plugins depended on others, that weren't target specific.

**Comment by jvdsande** [comment]
Hi,
You are totally correct, the underlying goal of all this is indeed a faster development mode. 

I guess defining two different pipelines, one for watch and one for build, would not be too constraining and would already be a nice start. 

It has the advantage of being transparent for current users : the "pipeline" entry and the "build" command would stay totally untouched, while a new "watch" command would look for a "watchPipeline" and fallback to "pipeline" if no watch pipeline is specified...
Does that make sense?

**Comment by jvdsande** [comment]
I guess it is basically the same as @ryaninvents proposal

**Comment by muhajirdev** [comment]
any progress on this?

**Comment by FredKSchott** [comment]
restarting this conversation a bit, I'd be interested in something similar to how Babel handles different "environments":

```
{
  "pipeline": [...], // all shared plugins
  "env": {
    "build": [...], // dev-build only plugins
    "publish": [...], // publish only plugins
    "test": [...], // user-defined environments, run via `--env test`
  }
```

**Comment by jvdsande** [comment]
Using this organisation, how would it account for the order of plugins?
If I'm not mistaken, the order is mandatory, e.g. the plugin-ts-standard-pkg comes first

**Comment by FredKSchott** [comment]
You're right that order is important, and I was picturing the shared "pipeline" running first, and then any "env" specific pipelines running after. 

I can't think of a real-world example of a package needing different "source" build plugins (the plugins that run earliest and create the "pkg/source" directory that all others build off of). I think that would be a safe assumption to make for the first version of this.

**Comment by jvdsande** [comment]
I see.
In my own experience, I usually swap _only_ the very first package, toggling between `standard-pkg` during my dev/watch steps, and `ts-standard-pkg` for production builds.

But I guess the same effect could be obtained by running `build-types` in production only at the end of the pipeline, and sticking with `standard-pkg` all the way
